src/Features/LanguageServer/Protocol/Protocol/Internal/VSInternalClientCapabilities.cs

```diff
@@ -8,11 +8,22 @@ namespace Roslyn.LanguageServer.Protocol
     using Newtonsoft.Json;
 
     /// <summary>
-    /// Extension class for ClientCapabilities with fields specific to Visual Studio.
+    /// Extension class for <see cref="ClientCapabilities"/> with fields specific to Visual Studio.
     /// </summary>
     [DataContract]
     internal class VSInternalClientCapabilities : ClientCapabilities
     {
+        /// <summary>
+        /// Gets or sets the text document capabilities from the Visual Studio client.
+        /// </summary>
+        [DataMember(Name = "textDocument")]
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public VSInternalTextDocumentClientCapabilities? VSInternalTextDocument
```

Comment:
I'm not seeing this on https://devdiv.visualstudio.com/DevDiv/_git/VSLanguageServerClient?path=/src/product/Protocol/LanguageServer.Protocol.Internal/VSInternalClientCapabilities.cs&_a=contents&version=GBdevelop
Why do we need it?

I would think it would also conflict with the `textDocument` field that is already on the base ClientCapabilities (when we serialize things)

Reply:
I suspect because VS doesn't actually support dynamic registration yet. I believe this works because use this type for the request handler and therefore the json serialization. Hiding the base TextDocument with `new` only works half way until someone uses just the base type, which would then return null. Although, thinking about that I could have set the base.TextDocument the same way as I do it now (originally I just had get; set;). Let me try that...

Reply:
OK, that worked. It makes it a bit clearer what the intent is - to change the type of the TextDocument property.

Reply:
Apologies - I might be being dense here, but I don't get why the type definition needs to be changed - the general pattern for the VS extensions onto a type is that the base definition is always used, but deserialization is then responsible for deserializing it into an instance of the VS type (if applicable).

So the `InitializeParams` would be the advertized type definition - but the instance recieved in the handler would be a `VSInternalInitializeParams`.  Is that not happening?

Reply:
@dibarbet That is happening. We get `VSInternalInitializeParams` but the `Capabilities` will still be `ClientCapabilities` and not `VSInternalClientCapabilities`. Basically, `VSInternalInitializeParams` doesn't do what it needs to do because VS does not support dynamic capabilities (yet). I'm just adding support for it in Roslyn now because we need it for VS Code.